ReSharper.FSharp/src/FSharp.Psi.Features/src/Parsing/FSharpTreeBuilderBase.fs

```diff
@@ -117,32 +117,21 @@ type FSharpTreeBuilderBase(lexer, document: IDocument, lifetime, path: VirtualFi
     member x.AdvanceToTokenOrRangeEnd(tokenType: TokenNodeType, range: range) =
         x.AdvanceToTokenOrPos(tokenType, range.End)
 
-    member x.AdvanceToXmlDoc(xmlDoc: XmlDoc, maxPos: pos) =
-        x.AdvanceToStart(xmlDoc.Range)
-
-        // todo: add separate token type for xmlDoc comment?
-        let maxOffset = x.GetOffset(maxPos)
-        while x.CurrentOffset < maxOffset && not (x.TokenType == FSharpTokenType.LINE_COMMENT && startsWith "///" (x.Builder.GetTokenText())) do
-            x.AdvanceLexer()
-
     member x.AdvanceToTokenOrPos(tokenType: TokenNodeType, pos: pos) =
         let maxOffset = x.GetOffset(pos)
         while x.CurrentOffset < maxOffset && x.TokenType != tokenType && not x.Eof do
             x.AdvanceLexer()
 
-    // todo: wrong ranges (e.g. missing union case bar)
-    // todo wrong order:
-    // todo multiple blocks:
     member x.MarkXmlDocOwner(xmlDoc: XmlDoc, expectedType: TokenNodeType, declarationRange: range) =
-        let xmlDocRange = xmlDoc.Range
-        let declarationStart = declarationRange.Start
-
-        if xmlDoc.IsEmpty || Position.posLt declarationStart xmlDocRange.Start then
-            x.AdvanceToTokenOrRangeStart(expectedType, declarationRange)
-        else
-            x.AdvanceToXmlDoc(xmlDoc, declarationStart)
+        x.AdvanceToTokenOrRangeStart(expectedType, declarationRange)
+        let mark = x.Mark()
+        if not xmlDoc.IsEmpty then
+            x.MarkAndDone(xmlDoc.Range, FSharpTokenType.XML_DOC_BLOCK)
+        mark
 
-        x.Mark()
+    member x.Mark(xmlDoc: XmlDoc, declarationRange: range) =
+        if xmlDoc.IsEmpty then x.Mark()
```

Comment:
Is this expected that this code must only be called when builder is at the declaration range start position?
Otherwise it may mark some random place prior to the declaration range if `xmlDoc` is empty.

We don't have this requirement in other `Mark*` methods that take ranges to mark, they advance the builder internally.

Reply:
This method is currently used only inside the processing of (top/local) bindings, where it is required that various keywords get inside. Perhaps we should inline its body to the call place.